R/stat-density-2d.r

```diff
@@ -60,12 +60,13 @@ StatDensity2d <- ggproto("StatDensity2d", Stat,
       lims = c(scales$x$dimension(), scales$y$dimension())
     )
     df <- data.frame(expand.grid(x = dens$x, y = dens$y), z = as.vector(dens$z))
+    df$nz <- df$z/max(df$z)
```

Comment:
Why `nz`? This naming seems strange.

Reply:
I wasn't sure what to call this one. I went with `nz` for "normalized z" as in `ndensity` and `ncount`. In my own code, I normally name variables which are derived from other variables using "x.difference" so `z` would become, `z.scaled` or similar, but I wasn't sure what the convention was for ggplot2.  

Alternatively, we could just call it `scaled`, but then the renaming that happens a few lines down would be somewhat redundant.

Reply:
I find it strange to introduce a variable name and then change it 5 lines down. Also, what happens if the `if (contour) {` branch is executed? What does `StatContour` do to the `nz` column?

Reply:
I'm inclined to agree; however, I tried to preserve as much of the approach of the original `stat_density` function as I could and that was the approach taken there:

```{r}
#exerpt from stat_denisty, line 62
    df <- data.frame(expand.grid(x = dens$x, y = dens$y), z = as.vector(dens$z))
    df$group <- data$group[1]

    if (contour) {
      StatContour$compute_panel(df, scales, bins, binwidth)
    } else {
      names(df) <- c("x", "y", "density", "group")
      df$level <- 1
      df$piece <- 1
      df
    }
```
`StatContour` doesn't do anything with the `nz` column, so to avoid having to rename things we could just name it when we create it.
```{r}
    df <- data.frame(expand.grid(x = dens$x, y = dens$y), z = as.vector(dens$z))
    df$scaled <- df$z / max(df$z)
    df$group <- data$group[1]

    if (contour) {
      StatContour$compute_panel(df, scales, bins, binwidth)
    } else {
      names(df) <- c("x", "y", "density", "scaled", "group")
      df$level <- 1
      df$piece <- 1
      df
    }
```
Or we could just create it later on, which avoids passing a column that won't be used to `StatContour`
```{r}
    df <- data.frame(expand.grid(x = dens$x, y = dens$y), z = as.vector(dens$z))
    df$group <- data$group[1]

    if (contour) {
      StatContour$compute_panel(df, scales, bins, binwidth)
    } else {
      names(df) <- c("x", "y", "density", "scaled", "group")
      df$scaled <- df$density / max(df$density)
      df$level <- 1
      df$piece <- 1
      df
    }
```
Or perhaps there's a third even cleaner method to achieve the same goal. Which approach do you think would be best?

Reply:
I think the later creation is better, though you shouldn't name it before you add it to the data frame. :-)

Reply:
Fixed! ✨ 

---
This may be a separate branch/pull request, but it might be worthwhile to change the behavior of `stat_density2d` when `aes(fill = stat(level))` is called when `contour = FALSE` and the behavior of `aes(fill = stat(density))` when `contour = TRUE` to return an error or a warning. Neither of those statistics are meaningful when called in the wrong context and it can return some cryptic errors as seen below. 😬 

```{r}
library(ggplot2)
library(dplyr)
library(viridis)
#2d densiy plot colored by level
diamonds %>%
ggplot(aes(x=x, y= depth)) +
  stat_density_2d(aes(fill = stat(density)),
                  geom = "raster",
                  n = 100,
                  bins = 10,
    contour = T) +
  facet_wrap(clarity~.) +
  scale_fill_viridis(option = "A")

#>  Error in is.finite(x) : default method not implemented for type 'closure' 
```
```{r}
diamonds %>%
ggplot(aes(x=x, y= depth)) +
  stat_density_2d(aes(fill = stat(level)),
                  geom = "polygon",
                  n = 100,
                  bins = 10,
    contour = F) +
  facet_wrap(clarity~.) +
  scale_fill_viridis(option = "A")
```
![plot7](https://user-images.githubusercontent.com/22528952/40866906-eb932514-65c5-11e8-9136-162964039046.png)